### ISSUE TITLE
wrong method is called (twig + doctrine)

### ISSUE THREAD

**Issue by susnux** [twigphp/Twig#3137]
Using twig and doctrine together I noticed, that twig calls the wrong function of an object.

Background: If you extra lazy load entities with doctrine, the count function issues just a COUNT() SQL query instead of a full SELECT() SQL query, this saves a lot of time.

Problem: Using a twig template like this:

``{% if book.pages.count > 0 %} something {% endif %}``

is not calling the `count()` function on the `book.pages` object, neither `count(book.pages)`, but it looks like it is calling the `count()` function on the parent of the pages object.

So I checked which class is returned when calling book.pages, it is the correct `Doctrine\\ORM\\PersistentCollection` and implements the correct `count()`). The parent class of this is `AbstractLazyCollection` which does not has this lazy load feature.

So to test this more, I tried this twig template versions (they *are* working), so I guess the method resolving of objects is some kind of broken:

``{% if book.pages.count() > 0 %} something {% endif %}``

and this:

``{% if book.pages|length > 0 %} something {% endif %}``

FYI, this is the doctrine part:

```
	/**
	 * @ORM\OneToMany(targetEntity="Page", mappedBy="book", fetch="EXTRA_LAZY")
	 */
	protected $events;
```

**Comment by stof** [comment]
well, Doctrine issues a `COUNT()` query in the PersistentCollection only if the collection is not initialized (otherwise, it would be a waste of DB query as all data is already in memory). Are you sure that your `{% if book.pages.count > 0 %}` is actually the place triggering the loading of the full collection (the `SELECT` query).

I do have similar cases in my own project, and they are working as expected.

**Comment by susnux** [comment]
@stof Yes, I have tested it, the only thing I have changed is the template (I use a SQL logger).
So `book.pages.count` triggers the full load, `book.pages.count()` does not, neither `book.pages | length`.

-----
**tl;dr**
Might be just a documentation bug if the `ArrayAccess` behavior using `object.element` is intended and for method access `object.method()` should be used.

**Long version:**
I have debugged it a bit more, the issue is inside of `twig_get_attribute`, because when calling with `count()` the type is set to `method`, this is then the correct member of PersistentCollection.
But when calling `.count` twig set type to `any` which then will use the first `if` branch (because PersistentCollection is of type `ArrayAccess`.

This leads to a call to `offsetExists` and this to `containsKey` which load the complete Collection.

So I am not so sure anymore that this is a twig bug, but more like incomplete documentation (as it might be intended to use `variable.element` when accessing a member variable and `variable.method()` when accessing a member method, but this is not clear in the twig documentation (the variables section does not even mention the possibility to use parenthesis to call a method).

So I leave this bug open, so you might decide if this bug is invalid, valid or just a documentation issue.

**Comment by stof** [comment]
ah, then indeed it makes sense. `book.pages.count` is meant to first try to read the property or ArrayAccess index, and then fallback to methods. This is the documented behavior of the `.` operator.
Earlier, I missed the fact that you were using an attribute access to get the count (that's the different with my project, where I always use `|length`)

I suggest you to use the `|length` filter to get the number of elements in your collection, which will use the Countable implementation directly.

**Comment by susnux** [comment]
If using `foo.bar()` is not encouraged I will close this, if it is supported and intended usage of twig I would add a section to the documentation similar to the  _accessing items on PHP arrays_ section (?).

**Comment by fabpot** [comment]
I'm closing as I don't think there is anything that is not already documented.
